File: Site_Web/covoiturage_la_cite_site_web/core/models/NotificationModel.ts
/**
 * NotificationModel — Modèle unifié et enrichi pour les notifications
 *
 * Chaque type peut embarquer un payload contextuel (tripDetails,
 * reservationDetails, reviewDetails…) pour afficher une vue détaillée
 * complète sans requête supplémentaire.
 */

// ─── Types de notification ────────────────────────────────────────────────────

export type NotificationType =
  | 'reservation_received'    // Conducteur : nouvelle demande de réservation
  | 'reservation_sent'        // Passager   : accusé de réception (envoi confirmé)
  | 'reservation_accepted'    // Passager   : demande acceptée par le conducteur
  | 'reservation_refused'     // Passager   : demande refusée par le conducteur
  | 'reservation_cancelled'   // Les deux   : réservation annulée
  | 'trip_created'            // Conducteur : trajet publié avec succès
  | 'trip_starting_soon'      // Les deux   : rappel de départ (< 1h)
  | 'trip_started'            // Passager   : conducteur a démarré le trajet
  | 'trip_completed'          // Les deux   : trajet terminé avec succès
  | 'trip_cancelled'          // Passager   : trajet annulé par le conducteur
  | 'new_review_received'     // Les deux   : nouvel avis reçu
  | 'cancellation_penalty'    // Les deux   : pénalité d'annulation appliquée
  | 'security_alert'          // Les deux   : alerte de sécurité (nouvelle connexion)
  | 'system';                 // Générique — avec ou sans lien attaché

// ─── Payloads contextuels ────────────────────────────────────────────────────

/** Détails d'un trajet embarqués dans la notification */
export interface NotificationTripDetails {
  tripId: string;
  departure: string;
  arrival: string;
  /** Format "YYYY-MM-DD" */
  date: string;
  /** Format "HH:mm" */
  time: string;
  /**
   * Prix affiché :
   * - passengerPrice (frais 15 % inclus) si le destinataire est un passager
   * - pricePerPassenger si le destinataire est le conducteur
   */
  price: number;
  availableSeats?: number;
  estimatedDurationMinutes?: number;
}

/** Détails d'une réservation embarqués dans la notification */
export interface NotificationReservationDetails {
  reservationId: string;
  /** Nom du passager demandeur (affiché chez le conducteur) */
  passengerName?: string;
  passengerAvatar?: string;
  passengerRating?: number;
  passengerTripCount?: number;
  /** Nom du conducteur (affiché chez le passager) */
  driverName?: string;
  driverAvatar?: string;
}

/** Détails d'un avis reçu */
export interface NotificationReviewDetails {
  reviewId: string;
  reviewerName: string;
  reviewerAvatar?: string;
  rating: number;
  /** Extrait du commentaire (120 chars max) */
  comment: string;
}

/** Détails d'une alerte de sécurité */
export interface NotificationSecurityDetails {
  clientType: 'web' | 'mobile';
  location?: string;
  userAgent?: string;
}

// ─── Modèle principal ─────────────────────────────────────────────────────────

export interface NotificationModel {
  // ── Identité ──────────────────────────────────────────────────────────────
  id: string;

  // ── Destinataire ──────────────────────────────────────────────────────────
  userId: string;

  // ── Contenu principal ─────────────────────────────────────────────────────
  type: NotificationType;
  /** Titre court affiché en gras dans l'alerte et la liste */
  title: string;
  /** Corps du message — lisible sans contexte supplémentaire */
  message: string;

  // ── État ──────────────────────────────────────────────────────────────────
  isRead: boolean;
  /** true → style urgent (rouge + pulse), priorité en tête de liste */
  isImportant: boolean;

  // ── Navigation ────────────────────────────────────────────────────────────
  /** URL de redirection au clic sur l'alerte ou le bouton d'action */
  link?: string;
  /** Libellé du bouton d'action dans la vue détail (dérivé du type si absent) */
  linkLabel?: string;

  // ── Relations ─────────────────────────────────────────────────────────────
  relatedTripId?: string;
  relatedReservationId?: string;

  // ── Payloads enrichis (vue détail sans requête supplémentaire) ────────────
  tripDetails?: NotificationTripDetails;
  reservationDetails?: NotificationReservationDetails;
  reviewDetails?: NotificationReviewDetails;
  securityDetails?: NotificationSecurityDetails;

  // ── Metadata ──────────────────────────────────────────────────────────────
  createdAt: string;
}

// ─── Constantes ───────────────────────────────────────────────────────────────

export const IMPORTANT_NOTIFICATION_TYPES: NotificationType[] = [
  'reservation_received',   // Conducteur : action requise
  'reservation_accepted',   // Passager   : bonne nouvelle urgente
  'trip_starting_soon',
  'trip_started',
  'trip_cancelled',
  'security_alert',
  'cancellation_penalty',
];

export function isImportantNotificationType(type: NotificationType): boolean {
  return IMPORTANT_NOTIFICATION_TYPES.includes(type);
}

File: App_Mobile/Covoiturage_la_cite_(App_Mobile)/Covoiturage_la_cite_(App_Mobile)/Core/Models/NotificationModel.cs
// ============================================================
//  Core/Models/NotificationModel.cs
//  Modèle unifié des notifications (miroir du web)
//  Payloads contextuels embarqués pour affichage sans requête.
// ============================================================

namespace Covoiturage_la_cite__App_Mobile_.Core.Models
{
    // ── Types de notification ────────────────────────────────
    public enum NotificationType
    {
        ReservationReceived,   // Conducteur : nouvelle demande
        ReservationSent,       // Passager   : accusé de réception
        ReservationAccepted,   // Passager   : demande acceptée
        ReservationRefused,    // Passager   : demande refusée
        ReservationCancelled,  // Les deux   : réservation annulée
        TripCreated,           // Conducteur : trajet publié
        TripStartingSoon,      // Les deux   : départ imminent
        TripStarted,           // Passager   : trajet démarré
        TripCompleted,         // Les deux   : trajet terminé
        TripCancelled,         // Passager   : trajet annulé
        NewReviewReceived,     // Les deux   : nouvel avis
        CancellationPenalty,   // Les deux   : pénalité appliquée
        SecurityAlert,         // Les deux   : nouvelle connexion
        System                 // Générique
    }

    // ── Payload : détails trajet ─────────────────────────────
    public record NotificationTripDetails(
        string TripId,
        string Departure,
        string Arrival,
        string Date,        // "YYYY-MM-DD"
        string Time,        // "HH:mm"
        double Price,
        int? AvailableSeats,
        int? EstimatedDurationMinutes
    );

    // ── Payload : détails réservation ───────────────────────
    public record NotificationReservationDetails(
        string ReservationId,
        string? PassengerName,
        string? PassengerAvatar,
        double? PassengerRating,
        int? PassengerTripCount,
        string? DriverName,
        string? DriverAvatar
    );

    // ── Payload : détails avis ───────────────────────────────
    public record NotificationReviewDetails(
        string ReviewId,
        string ReviewerName,
        string? ReviewerAvatar,
        double Rating,
        string Comment
    );

    // ── Payload : alerte sécurité ────────────────────────────
    public record NotificationSecurityDetails(
        string ClientType,    // "web" | "mobile"
        string? Location,
        string? UserAgent
    );

    // ── Modèle principal ─────────────────────────────────────
    public class NotificationModel
    {
        public string Id { get; init; } = "";
        public string UserId { get; init; } = "";

        public NotificationType Type { get; init; }
        public string Title { get; init; } = "";
        public string Message { get; init; } = "";

        public bool IsRead { get; set; }
        public bool IsImportant { get; init; }

        public string? Link { get; init; }
        public string? LinkLabel { get; init; }

        public string? RelatedTripId { get; init; }
        public string? RelatedReservationId { get; init; }

        // Payloads contextuels enrichis
        public NotificationTripDetails? TripDetails { get; init; }
        public NotificationReservationDetails? ReservationDetails { get; init; }
        public NotificationReviewDetails? ReviewDetails { get; init; }
        public NotificationSecurityDetails? SecurityDetails { get; init; }

        public string CreatedAt { get; init; } = "";
    }

    // ── Constantes ───────────────────────────────────────────
    public static class NotificationConstants
    {
        public static readonly IReadOnlySet<NotificationType> ImportantTypes =
            new HashSet<NotificationType>
            {
                NotificationType.ReservationReceived,
                NotificationType.ReservationAccepted,
                NotificationType.TripStartingSoon,
                NotificationType.TripStarted,
                NotificationType.TripCancelled,
                NotificationType.SecurityAlert,
                NotificationType.CancellationPenalty,
            };

        public static bool IsImportant(NotificationType type)
            => ImportantTypes.Contains(type);
    }
}
